pkg/yum/repository.go

```diff
@@ -0,0 +1,269 @@
+package yum
+
+import (
+	"compress/gzip"
+	"encoding/xml"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"net/url"
+	"path"
+)
+
+// Package metadata of a given package
+type Package struct {
+	Type     string   `xml:"type,attr"`
+	Name     string   `xml:"name"`
+	Arch     string   `xml:"arch"`
+	Version  Version  `xml:"version"`
+	Checksum Checksum `xml:"checksum"`
+	Summary  string   `xml:"summary"`
+}
+
+type Version struct {
+	Version string `xml:"ver,attr"`
+	Release string `xml:"rel,attr"`
+	Epoch   int32  `xml:"epoch,attr"`
+}
+
+type Checksum struct {
+	Value string `xml:",chardata"`
+	Type  string `xml:"type,attr"`
+}
+
+// Repomd metadata of the repomd of a repository
+type Repomd struct {
+	XMLName  xml.Name `xml:"repomd"`
+	Data     []Data   `xml:"data"`
+	Revision string   `xml:"revision"`
+}
+
+func (r *Repomd) Raw() ([]byte, error) {
+	return xml.Marshal(r)
+}
+
+type Data struct {
+	Type     string   `xml:"type,attr"`
+	Location Location `xml:"location"`
+}
+type Location struct {
+	Href string `xml:"href,attr"`
+}
+
+type YumRepository interface {
+	FetchRepomd() (statusCode int, err error)
+	FetchPackages() (statusCode int, err error)
+	FetchSignature() (statusCode int, err error)
+	Repository() Repository
+}
+
+type Repository struct {
+	Client    http.Client // Client to fetch repository data
+	URL       string      // URL of the repository
+	Packages  []Package   // Packages repository contains
+	GPGKey    *string     // GPG key of the repository
+	Signature *string     // Signature of the repository
+	Repomd    Repomd      // Repomd of the repository
+}
+
+func (r *Repository) Repository() Repository {
```

Comment:
I had to add this method to as a work around to be able to mock in the backend. Not sure I like this.